### DISCUSSION TITLE
How to change the Prediction Horizon during at runtime

### DISCUSSION THREAD

**Discussion by MikAmb95** [casadi/casadi#3587] [Q&A] [answered]
Hello!!!! 

I'm trying to implement a NLMPC where the prediction horizon is not fixed but it can change at runtime. 

The only solution that I found so far is to re-build the solver() every time that I need to change the prediction horizon. But it takes a lot of time.

Is there another way to do it more efficiently?

Thank you!!

**Comment by brunoml5** [comment]
One thing you can do is to change variables so that time is a parameter, or a state. Take a look at this [reply](https://github.com/casadi/casadi/discussions/3582#discussioncomment-8452903) from me, see if it helps.

Another option is to play with constraints and objective function. You could add parameters on the continuity constraints of your MPC so that you can activate or deactivate them at will. 

There might be other options. Or even toolboxes that do it for your. Hopefully somebody comes and complements my reply.

**Reply by MikAmb95** [reply]
Hi! Thanks for your answer. 

I tried playing with constraints and the obj function as you suggested. But every time I change constraints/obj function it seems I have to rebuild the solver. Assuming that at the beginning my NMPC has a prediction horizon N. If I want to reduce or change it this is the code that I'm using
```
if flag == true
N = N -1;  

nlp_prob = struct('f', obj, 'x', OPT_variables, 'g', g, 'p', P);  % after modifying the constraints and the objective function and decreased the number of variables.

solver = nlpsol('solver', 'ipopt', nlp_prob,opts); % re-build solver THIS TAKES TIME

end

```
 
Is there any way to avoid rebuilding the solver every time I reduce N? 

thanks!!!

**Reply by brunoml5** [reply]
No no, the way you are doing, you are creating a solver every time. I meant something like this (syntax based on direct_multiple_shooting.py):
```
# Create a auxiliary parameter for your NLP
aux_N = MX.sym('aux_N', 1)
p += [aux_N]

# Add equality constraint (continuity equation)
g   += [(Xk_end-Xk)*aux_N]
lbg += [0, 0]
ubg += [0, 0]

prob = {'f': J, 'x': vertcat(*w), 'g': vertcat(*g), 'p': vertcat(*p)}
solver = nlpsol('solver', 'ipopt', prob);
```
This way, when you call your solver with `aux_N=0`, then your continuity equation will be zero regardless of `Xk`, so it should not affect the result of the rest of your NLP. And if it is `aux_N=1`, then the constraint is active. Maybe do the same for your stage cost in your objective function.

PS: this is just a sketch so that you understand the concept. If you would like a full example, tell me.

**Reply by MikAmb95** [reply]
Hi! Thanks again for your answer! 

I get the idea, if you could provide a complete example that would be great.  

However, with your suggested solution, the number of opt variable remains the same. Right? While instead if the prediction horizon is reduced, the number of variables should also be lowered.  

Thanks for your help!

**Reply by brunoml5** [reply, marked as answer]
I don't see how you can reduce the number of optimization variables without creating additional solvers. If you really need that, you could create several solver for each horizon lenght offline, then select them online.

The full example based on direct_multiple_shooting.py:
```
from casadi import *

T = 10. # Time horizon
N = 20 # number of control intervals

# Declare model variables
x1 = MX.sym('x1')
x2 = MX.sym('x2')
x = vertcat(x1, x2)
u = MX.sym('u')

xdot = vertcat((1-x2**2)*x1 - x2 + u, x1)   # Model equations
L = x1**2 + x2**2 + u**2    # Objective term

# Fixed step Runge-Kutta 4 integrator
M = 4       # RK4 steps per interval
DT = T/N/M
f = Function('f', [x, u], [xdot, L])
X0 = MX.sym('X0', 2)
U = MX.sym('U')
X = X0
Q = 0
for j in range(M):
    k1, k1_q = f(X, U)
    k2, k2_q = f(X + DT/2 * k1, U)
    k3, k3_q = f(X + DT/2 * k2, U)
    k4, k4_q = f(X + DT * k3, U)
    X=X+DT/6*(k1 +2*k2 +2*k3 +k4)
    Q = Q + DT/6*(k1_q + 2*k2_q + 2*k3_q + k4_q)
F = Function('F', [X0, U], [X, Q],['x0','p'],['xf','qf'])

# Start with an empty NLP
w, w0, lbw, ubw = [], [], [], []
g, lbg, ubg = [], [], []
p = []
J = 0

# "Lift" initial conditions
Xk = MX.sym('X0', 2)
w += [Xk]
lbw += [0, 1]
ubw += [0, 1]
w0 += [0, 1]

aux_N = MX.sym('aux_N', N)
p += [aux_N]

# Formulate the NLP
for k in range(N):
    # New NLP variable for the control
    Uk = MX.sym('U_' + str(k))
    w += [Uk]
    lbw += [-1]
    ubw += [1]
    w0 += [0]

    # Integrate till the end of the interval
    Fk = F(x0=Xk, p=Uk)
    Xk_end = Fk['xf']

    J = J + Fk['qf']*aux_N[k]   # Multiply the objective term by aux_N[k]

    # New NLP variable for state at end of interval
    Xk = MX.sym('X_' + str(k+1), 2)
    w += [Xk]
    lbw += [-0.25, -inf]
    ubw += [  inf,  inf]
    w0 += [0, 0]

    # Add equality constraint (continuity equation) with the term aux_N[k]
    g   += [(Xk_end-Xk)*aux_N[k]]
    lbg += [0, 0]
    ubg += [0, 0]

# Create an NLP solver
prob = {'f': J, 'x': vertcat(*w), 'g': vertcat(*g), 'p': vertcat(*p)}
solver = nlpsol('solver', 'ipopt', prob);

# Here you choose the length of the horizon which is active
active_N = 8
p_opt = [1]*active_N + [0]*(N-active_N)

# Solve the NLP
sol = solver(x0=w0, lbx=lbw, ubx=ubw, lbg=lbg, ubg=ubg, p=p_opt)
w_opt = sol['x'].full().flatten()

# Plot the solution
x1_opt = w_opt[0::3]
x2_opt = w_opt[1::3]
u_opt = w_opt[2::3]

tgrid = [T/N*k for k in range(N+1)]
import matplotlib.pyplot as plt
plt.figure(1)
plt.clf()
plt.plot(tgrid, x1_opt, '--')
plt.plot(tgrid, x2_opt, '-')
plt.step(tgrid, vertcat(DM.nan(1), u_opt), '-.')
plt.xlabel('t')
plt.legend(['x1','x2','u'])
plt.grid()
plt.show()
```
You will notice that the variables outside the active horizon results in weird variables.

**Comment by MikAmb95** [comment]
Hi thanks a lot for your time and help!! 

I tried the two solutions. 1) create several solvers for each horizon length offline 2) the proposed solution in your last answer. Both works well, but the solution 1) is faster for a long initial prediction horizon. 

thank you again for your help!

**Reply by brunoml5** [reply]
If you are using several solvers, make sure you are warm starting it properly. It helps a lot with convergence.

See the documentation for ipopt, it also needs the dual variables. And for online MPC, consider an SQP method because it better exploits warm start.

**Reply by MikAmb95** [reply]
I will definitely try to improve the code by following your suggestions. thanks again for your help and time.
